src/dune_tui/dune_tui.ml

```diff
@@ -144,35 +33,291 @@ module Tui = struct
     fun () -> !(Lazy.force setup)
   ;;
 
-  let start () = ()
+  (** style for diving visual elements like borders or rules *)
+  let divider_attr = A.(fg red)
+
+  (** style for helpful ui elements like scrollbar structures or help text *)
+  let helper_attr = A.(fg yellow)
+
+  (** style for user feedback like message count, or scrollbar position *)
+  let user_feedback_attr = A.(fg cyan)
+
+  (** [horizontal_rule ~attr ~w] draws a horizontal line with [attr] of length [w]. *)
+  let horiztonal_rule ~attr ~w = I.uchar attr (Uchar.of_int 0x2015) w 1
+
+  (** Here we keep some persistent state about the program that we udpate each time we
+      render. This allows other components to "react" to changes using [Lwd]. *)
+
+  let term_size = Lwd.var (Term.size (term ()))
+  let messages = Lwd.var []
+  let status_line = Lwd.var None
+
+  module Message_viewer = struct
+    (** Specialized widget for viewing messages in a list. *)
+
+    (** [is_message_hidden] tracks if a given message is hidden. [true] means it is hidden
+        and [false] means it is expanded. *)
+    let is_message_hidden = Lwd.var (fun _ -> false)
 
-  let image ~status_line ~messages =
+    (** We share some work turning messages into images. *)
+    let message_images =
+      let+ messages = Lwd.get messages in
+      List.map messages ~f:(fun x -> User_message_to_image.pp (User_message.pp x))
+    ;;
+
+    (** We calculate the max message length from all the messages. This is used to keep
+        the horizontal lines with a consistent length. *)
+    let max_message_length =
+      let+ messages = message_images
+      and+ w, _ = Lwd.get term_size in
+      let lengths = List.map messages ~f:I.width in
+      match List.max lengths ~f:Int.compare with
+      | None -> w - 1
+      | Some l -> max l w
+    ;;
+
+    (** We approximate the first line of the message. Unfortunately due to the way Notty
+        images work, it is not easy to get the actual width of the first line. Therefore
+        we just chop a third off as an approximation. *)
+    let message_synopsis ~attr =
+      let+ messages = message_images in
+      fun index ->
+        match List.nth messages index with
+        | None -> I.string attr "..."
+        | Some message ->
+          let cropped_image = I.vcrop 0 (I.height message - 1) message in
+          I.hcrop 0 (I.width cropped_image / 3) cropped_image
+    ;;
+
+    (* This is a line that shows the total number of messages and the message count used
+       for separating messages. It also has a handler for clicks that minimizes the
+       message. *)
+    let horizontal_line_with_count total index =
+      let+ is_hidden = Lwd.get is_message_hidden
+      and+ w = max_message_length
+      and+ synopsis = Lwd.app (message_synopsis ~attr:helper_attr) (Lwd.return index) in
+      let index_is_hidden = is_hidden index in
+      let status =
+        I.hcat
+          [ I.uchar divider_attr (Uchar.of_int 0x169c) 1 1
```

Comment:
Can you make all `Uchar.of_int <magic-number>` into proper constants and add some comments for them?